### DISCUSSION TITLE
Return isotile grid as raster from `/isochrone`

### DISCUSSION THREAD

**Discussion by chrstnbwnkl** [valhalla/valhalla#4581] [Ideas]
I recently thought it could be a good addition to offer the possibility to return the isotile grid in some way or another, instead of the contours that are formed from it. 

This is an example output of a quick and dirty implementation that uses GDAL to write a GeoTIFF:

![image](https://github.com/valhalla/valhalla/assets/58629404/b605a23a-8463-4c75-8553-e8397bf8f7d8)

There's no particular reason I chose GeoTIFF specifically except for 
  1. it's geospatial
  2. it's widely supported

And the only reason I chose GDAL was just pure lazyness :smile: But if this feature is something that people desire, I'd be willing to spend some time to polish up what I have. The output format doesn't need to be a standardized format of course, I could also imagine returning e.g. a protobuf message that contains the pixel array and some information about its origin and pixel size. If adding GDAL as a dependency is an issue, maybe it's not so difficult to write a simple GeoTIFF serializer (haven't looked into this).

Looking forward to hearing your thoughts!

**Comment by nilsnolde** [comment]
I love it! IMO it's even more meaningful than contours, which can even be generated in QGIS/other software from this grid (as you noted earlier). No idea about our own GeoTIFF implementation either, I could imagine it's quite a bit of work. 

I'm a bit reserved about adding GDAL as runtime dependency, although it's surely the most stable (and available) geo library out there. It's just quite huge and monolithic for a single driver.. If it's a serious concern, we could also guard this with a `ENABLE_GDAL` compiler option, if people don't want this/GDAL. Should be easy enough, the code paths are very contained.

**Reply by nilsnolde** [reply]
Also would need a tiny addition in prime_server to allow for a `Content-type: application: image/tiff; application=geotiff`, and probably https://github.com/kevinkreiser/prime_server/issues/129 might benefit here. How big is the response for the above image?

**Reply by chrstnbwnkl** [reply]
The example above is a 1347x849 grid, which is a bit over 1mb using 8 bit ints, or 2 mb with 16 bit ints (uncompressed)

**Reply by nilsnolde** [reply]
Yeah fair:) Brings up the resolution of the metric. We'd need 16 bits quite sure. For time that'd make 18 hours at 1 second resolution, so we don't need to worry about running out. For distance it's a bit harder, 65536 meters is not much. So we could scale by factor 10, i.e. 10 m resolution for 655 km? Making resolution dynamic is not an option I think; that would work fine for anything with <= 16 bit max metric, but hard to communicate to the user for > 16 bit max metric when we need to scale. 10 m resolution is also acceptable IMO.

**Reply by chrstnbwnkl** [reply]
`distance` is in km though isn't it? That's how it's documented: https://valhalla.github.io/valhalla/api/isochrone/api-reference/

**Reply by nilsnolde** [reply]
right, but that's when we have floats. since we're dealing with integers here, I think providing meters is more intuitive. honestly, I don't really like that we're returning kilometers and minutes in floats over meters and seconds in integers. but that's a individual preference call I guess..

**Comment by kevinkreiser** [comment]
@dnesbitt61 always used to mention this as an idea iirc

prime server doesn't stop you from returning whatever content type you like. also gdals geotiff driver supports compression natively in its container. better to just use that. I don't think it makes sense to compress it again for the http transport as it would make little added difference

**Comment by kevinkreiser** [comment]
also tiff in general supports adding metadata tags so we can send some info about units resolution there if that could help

**Reply by chrstnbwnkl** [reply]
this could even be used to include color information, similar to how it's done on the json response

**Reply by nilsnolde** [reply]
ah didn't know there's metadata support, great!
